sdk/python/src/dagger/server/_converter.py

```diff
@@ -0,0 +1,107 @@
+from collections.abc import Callable
+from dataclasses import fields
+from typing import Annotated, TypeGuard, TypeVar, get_args, get_origin
+
+import anyio
+from cattrs import Converter
+from cattrs.gen import make_dict_structure_fn, make_dict_unstructure_fn, override
+from cattrs.preconf.json import make_converter
+
+from dagger.api.base import IDType, Scalar, Type
+
+from ._context import Context
+from ._util import has_resolver, is_strawberry_type
+
+T = TypeVar("T")
+
+converter = make_converter(omit_if_default=True)
+
+
+def strawberry_structure_hook(cls: type) -> Callable:
+    """Don't structure fields with resolvers."""
+    return make_dict_structure_fn(
+        cls,
+        converter,
+        **{
+            f.name: override(omit_if_default=True, omit=has_resolver(f))
+            for f in fields(cls)
+        },
+    )
+
+
+def strawberry_unstructure_hook(cls: type) -> Callable:
+    """Don't unstructure with resolvers."""
+    return make_dict_unstructure_fn(
+        cls,
+        converter,
+        **{
+            f.name: override(omit_if_default=True, omit=has_resolver(f))
+            for f in fields(cls)
+        },
+    )
+
+
+def is_dagger_id_type(cls: type) -> TypeGuard[IDType]:
+    return issubclass(cls, Type) and hasattr(cls, "id")
+
+
+def register_dagger_type_hooks(conv: Converter, ctx: Context):
+    """Register structure and unstructure hooks for dagger types.
+
+    Needs a factory function because of async client access.
+    """
+    from dagger.api import gen
+
+    # TODO: This is a hack. We need to add functions with codegen for
+    # dynamic situations like these.
```

Comment:
Yeah you probably already saw it, but I added this to the go codegen to help out with what I think is the same (or similar-ish) problem, just in case that's helpful: https://github.com/dagger/dagger/pull/5157/files#diff-31244223d8a860c03288b53863173f3e261164aa5a30c6656810826206939c39R19

Reply:
Yep. 😄 I noticed you had the same kind of TODO comment after I added it. Planning on doing it in this PR, just had to focus on other things for 2 days. Also noticed your fix 👍